File: Day272/parallelCoursesIII.java
package Day272;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.Queue;

public class parallelCoursesIII {

    public static int minimumTime(int n, int relations[][], int time[]) {
        int answer = 0;
        Queue<Integer> queue = new LinkedList<>();
        Map<Integer, List<Integer>> graph = new HashMap();
        int id[] = new int[n+1];
        int temp[] = new int[n+1];
        System.arraycopy(time, 0, temp, 1, n);
        for (int[] i: relations) {
           graph.computeIfAbsent(i[0], k-> new ArrayList<>()).add(i[1]);
           id[i[1]]++;
        }
        for (int i=1; i<= n; i++) {
            if (id[i] == 0 ) {
                queue.offer(i);
            }
        }
        while (!queue.isEmpty()) {
            int t = queue.poll();
            if (graph.containsKey(t)) {
                for (int nb : graph.get(t)) {
                    temp[nb] = Math.max(temp[nb], temp[t] + time[nb-1]);
                    id[nb]--;
                    if (id[nb] == 0) {
                        queue.offer(nb);
                    }
                }
            }
        }
        for (int i=0; i<=n; i++) {
            answer = Math.max(answer, temp[i]);
        }
        return answer;
    }

    public static void main(String[] args) {
        int n1 = 3;
        int relations1[][] = {{1,3}, {2,3}};
        int time1[] = {3,2,5};
        int n2 = 5;
        int relations2[][] = {{1,5}, {2,5}, {3,5}, {3,4}, {4,5}};
        int time2[] = {1,2,3,4,5};
        System.out.println(minimumTime(n1, relations1, time1));
        System.out.println(minimumTime(n2, relations2, time2));

    }
}

File: Day272/parallelCoursesIII.go
package main

import (
	"fmt"
	"math"
)

func minimumTime(n int, relations [][]int, time []int) int {
	graph := make(map[int][]int )
	queue := make([]int,0)
	id := make([]int, n+1)
	for _,v := range relations {
		graph[v[0]] = append(graph[v[0]], v[1])
		id[v[1]]++
	}
	temp := make([]int, n+1)
	copy(temp[1:], time)
	for i:=1; i<=n; i++ {
		if id[i] == 0 {
			queue = append(queue, i)
		}
	}
	for len(queue) > 0 {
		t := queue[0]
		queue= queue[1:]
		for _, neighbour := range graph[t] {
			temp[neighbour] =int(math.Max(float64(temp[neighbour]), float64(temp[t]+time[neighbour-1])))
			id[neighbour]--
			if id[neighbour] == 0 {
				queue = append(queue, neighbour)
			}
		}
	}
	max  := 0
	for _,v := range temp {
		max = int(math.Max(float64(max), float64(v)))
	}
	return max
}

func main() {
	n1 := 3
	relations1 := [][]int{{1,3}, {2,3}}
	time1 := []int{3,2,5}
	n2 := 5
	relations2 := [][]int{{1,5}, {2,5}, {3,5}, {3,4}, {4,5}}
	time2 := []int{1,2,3,4,5}
	fmt.Println(minimumTime(n1, relations1, time1))
	fmt.Println(minimumTime(n2, relations2, time2))
}